src/GafferScene/Instancer.cpp

```diff
@@ -2650,39 +2650,116 @@ void Instancer::InstancerCapsule::render( IECoreScenePreview::Renderer *renderer
+					// we can reuse the context. But timeOffset is relative, so it's important that we reset the
+					// time before we do setPrototypeContextVariables for the next element. ( Should this be more
+					// general instead of assuming that frame is the only variable for which offsetMode may be set? )
+					prototypeScope.setFrame( onFrameTime );
+
+					engines[0]->setPrototypeContextVariables( pointIndex, prototypeScope );
+
+					MurmurHash contextHash = prototypeScope.context()->hash();
+
+					// We only need a read lock to check if there is already a prototype corresponding to this
+					// context.
+					tbb::spin_rw_mutex::scoped_lock prototypeContextsLock( prototypeContextsMapMutex, false );
+					size_t prototypeIndex;
+					auto existingPrototypeIndex = prototypeContextsMap.find( contextHash );
+					if( existingPrototypeIndex != prototypeContextsMap.end() )
+					{
+						// Found an existing prototype, we just use this index
+						prototypeIndex = existingPrototypeIndex->second;
+						prototypeContextsLock.release();
+					}
+					else
+					{
+						// Need to add a new prototype, so we need a write lock on the contexts map to
+						// add a new index
+						prototypeContextsLock.upgrade_to_writer();
+						auto inserted = prototypeContextsMap.emplace( contextHash, prototypeContextsMap.size() );
 
-			prototypeIndices.push_back( inserted.first->second );
+						prototypeIndex = inserted.first->second;
 
-		}
+						// We now have an index, and no longer need to access the contexts map.
+						prototypeContextsLock.release();
+
+						if( !inserted.second )
+						{
+							// Someone else allocated a prototype for this context while we were acquiring the
+							// write lock, we don't need to add anything after all
+						}
+						else
+						{
+							// We're actually adding a prototype, so we need the `prototypesLock`
+							tbb::spin_rw_mutex::scoped_lock prototypesLock( prototypesMutex, false );
+
+							if( prototypes.size() <= prototypeIndex )
+							{
+								// We need to resize the `prototypes` vector, this is the only time
+								// we need a write lock.
+								prototypesLock.upgrade_to_writer();
+
+								// Double check that someone else hasn't resized it for us while we were getting
+								// the lock
+								if( prototypes.size() <= prototypeIndex )
+								{
+									// Resize up by a factor of 2, to amortize how many resizes we need to do
+									prototypes.resize( ( prototypeIndex + 1 ) * 2 );
+								}
+								prototypesLock.downgrade_to_reader();
+							}
+
+							// We've gotten a unique prototypeIndex from the `prototypeContextsMap`, and each
+							// element of `prototypes` can be set independently, so we can safely make this
+							// assignment with just a read lock on prototypesMutex
+
+							prototypes[prototypeIndex] = addPrototypeToVectors(
+									prototypesPlug, prototypeRoot, prototypeScope, sampleTimes, capsuleHash, *renderOptions
+							);
```

Comment:
This is likely where the greatest computational costs are in this loop, particularly in the not-a-capsule case where we're generating the object itself directly. And because we're holding the `prototypesLock`, we're blocking any thread needing to do a resize for the whole time we're generating the object. We're also unable to take any advantage of the waiting threads if there is parallelism in the creation of the prototype itself (for instance, an OSLObject deformer). And it seems fairly likely that we lose out on performance if we're left waiting for one thread to finish the last prototype before we can go on and process the first instance.

For the sake of argument, I think my instinct would have been to not have two loops, but to just have the making-the-instances loop, and generate the needed prototypes lazily within that. That could be done by storing the prototypes in an LRUCache, indexed by `contextHash` and sized such that we never evicted. If this used the `TaskParallel` policy, points wanting the same prototype would naturally collaborate on the inner parallelism (the hypothetical deformer) simply by virtue of calling `LRUCache::get()`. I also tend to think that having only a single loop might also be a bit easier to follow. Might be worth considering?

Reply:
I think you're probably going to be happier with the code structure now that I've switched to LRUCache, and the performance isn't worse, so it does seem like an overall win.

But at least one of the benefits you were hoping for didn't pan out: the overhead of `TaskParallel` resulted in a more than 2X slowdown on `testInstancerManyPrototypesEncapsulatePerf`. Switching to just `Parallel` left me about where I was before - so I guess this is another good test case if you want to look into reducing tbb overhead.